prbt_hardware_support/test/integrationtests/integrationtest_speed_observer_and_operation_mode.cpp

```diff
@@ -0,0 +1,290 @@
+  nh_.shutdown();
+}
+
+void SpeedObserverIntegrationTest::publishOm(OperationModes::_value_type omv){
+  OperationModes om;
+  om.value = omv;
+  om.time_stamp = ros::Time::now();
+  operation_mode_pub_.publish(om);
+}
+
+void SpeedObserverIntegrationTest::publishJointStatesAtSpeed(double v)
+{
+  sensor_msgs::JointState js;
+  js.name = { "testing_world-a" };
+  ros::Rate r = ros::Rate(TEST_FREQUENCY * 3);  // publishing definitely faster
+                                                // then observing
+  ros::Time start = ros::Time::now();
+  double t = 0;
+  joint_publisher_running_ = true;
+  while (joint_publisher_running_ & nh_.ok())
+  {
+    ros::Time current = ros::Time::now();
+    t = (current - start).toSec();
+    js.position = { t * v };
+
+    fake_controller_joint_states_pub_.publish(js);
+    if (joint_publisher_running_ & nh_.ok())  // ending fasteroperation_mode_cb
+      r.sleep();
+  }
+}
+
+void SpeedObserverIntegrationTest::stopJointStatePublisher()
+{
+  joint_publisher_running_ = false;
+}
+
+/**
+ * @tests{Stop1_on_violation_of_speed_limit}
+ *
+ * Test Sequence:
+ *    0. Set Operation Mode to T1 by answering the service call
+ *    1. Joint motion with a speed higher than the T1 limit is faked.
+ *    2. Set Operation Mode to Auto by publihsing on the topic
+ *       Wait for stop calls that may occurr from previous run
+ *    3. Joint motion with a speed lower than the Auto limit is faked.
+ *    4. Switching back to T1
+ *    5. Set up the STO service to return success = false
+ *
+ * Expected Results:
+ *    0. -
+ *    1. Operation mode not to be called any more
+ *       Sto is triggered
+ *    2. -
+ *    3. Operation mode not to be called any more
+ *       Sto is *not* triggered
+ *    4. Sto is triggered
+ *    5. Sto is triggered and error message is produced
+ */
+TEST_F(SpeedObserverIntegrationTest, testT1)
+{
+  Sequence s_speeds, s_stop;
+  /**********
+   * Step 0 *
+   **********/
+  ROS_DEBUG("Step 0");
+
+  // Set OM to T1
+  GetOperationMode::Response om_res;
+  om_res.mode.value = OperationModes::T1;
+  om_res.mode.time_stamp = ros::Time::now();
+  EXPECT_CALL(*this, operation_mode_cb(_, _)).InSequence(s_stop)
+      .WillOnce(DoAll(SetArgReferee<1>(om_res), ACTION_OPEN_BARRIER(BARRIER_OPERATION_MODE_SET)));
+  EXPECT_CALL(*this, frame_speeds_cb(ContainsAllNames(additional_frames_))).Times(AtLeast(1)).InSequence(s_speeds);
+  EXPECT_CALL(*this, sto_cb(_, _)).Times(0);
+
+  BARRIER({ BARRIER_OPERATION_MODE_SET });
+
+  /**********
+   * Step 1 *
+   **********/
+  ROS_DEBUG("Step 1");
+
+  // Now the service should not be called any more
+  EXPECT_CALL(*this, operation_mode_cb(_, _)).Times(0).InSequence(s_stop);
+  EXPECT_CALL(*this, sto_cb(StoState(true), _))
+      .WillRepeatedly(DoAll(SetArgReferee<1>(sto_res), ACTION_OPEN_BARRIER(BARRIER_STOP_HAPPENED)));
+
+  std::thread pubisher_thread = std::thread(&SpeedObserverIntegrationTest::publishJointStatesAtSpeed, this, 1.0);
```

Comment:
How can u be sure that the Limit was triggered at the configured frame speed?

Reply:
what do you mean?

Reply:
we cannot be sure.